kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala

```diff
@@ -83,7 +85,9 @@ private[kyuubi] class EngineRef(
   // Launcher of the engine
   private[kyuubi] val appUser: String = shareLevel match {
     case SERVER => Utils.currentUser
-    case GROUP =>
+    case GROUP => if (userGroupMap.nonEmpty) {
+        userGroupMap.getOrElse(user, user)
+      } else {
```

Comment:
we actually need a `UserGroupProvider` here, and the implementation could be `HadoopUserGroupProvider`, `UnixUserGroupProvider`, `LDAPUserGroupProvider` `JbdcUserGroupProvider` etc. But TBH, we should be careful about providing the JDBC-based implementation too early, because once done, we have responsibilities to keep compatibility and database upgrading scripts.

Reply:
I try my best to recall some previous points on the "group".

- one user may have multi groups, so the question here are,
  - only use the primary group, or use the primary group in default but allow the user to use the specific one?
  - all groups should be passed to the engine, it's useful for cases like authz plugin.

Further, some admins want the user to use their personal account(e.g. LDAP) for authentication, but run Job in service account, it's also a kind of "user group mapping" case IMO.

Reply:
cc @leoluan2009 as you are working on LDAP enhancement, this is something we need to do in the next step.

Reply:
Thank you for your valuable advice，let me soak it up.

Reply:
FYI, the pluggable `GroupProvider` is added in https://github.com/apache/incubator-kyuubi/pull/3897

Reply:
Ok, thank you